imagecraft/grammar.py

```diff
@@ -0,0 +1,114 @@
+# This file is part of imagecraft.
+#
+# Copyright 2025 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License version 3, as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Grammar processor."""
+
+from typing import Any, cast
+
+import craft_cli
+from craft_application.errors import CraftValidationError
+from craft_grammar import GrammarProcessor  # type: ignore[import-untyped]
+from craft_grammar.errors import GrammarSyntaxError  # type: ignore[import-untyped]
+
+from imagecraft.models import get_grammar_aware_volume_keywords
+
+# Values that should return as a single object / list / dict.
+_NON_SCALAR_VALUES = [
+    "structure",
+]
+
+# Values that should return a dict, not in a list.
+_DICT_ONLY_VALUES: list[str] = []
+
+
+def process_volumes(
+    *, volumes_yaml_data: dict[str, Any], arch: str, target_arch: str
+) -> dict[str, Any]:
+    """Process grammar for volumes.
+
+    :param yaml_data: unprocessed volumes section of imagecraft.yaml.
+    :returns: processed volumes section of imagecraft.yaml.
+    """
+
+    def self_check(value: Any) -> bool:  # noqa: ANN401
+        return bool(
+            value == value  # pylint: disable=comparison-with-itself  # noqa: PLR0124
```

Comment:
I don't understand this in context. The only non-true value I can think of is NaN, but I'm not sure how we'd get that?

Reply:
Out of coherence I used the same implementation of `self_check` as in snapcraft and craft-application. I was very tempted to simply `return True` but I suspect there are corner cases I might have missed.

I doubt we could end up with a NaN value since that would be caught by the model validation, but I confirm that would return `False` (I tested with `a = float('nan'); a ==a`)